[SRVKE-897] Improve monitoring reconciliation for sources

skonto opened the pull request:
- Fixes SRVKE-897
- Makes monitoring setup for sources easier to configure. Right now by default we setup cluster monitoring for sources per namespace they reside in, for better get started UX. We do this by marking the corresponding ns with a special flag: `openshift.io/cluster-monitoring=true`. This causes any services running side by side with a source (same namespace) being scraped by the cluster monitoring Prometheus instance. This should be avoided in general because user services should have a different metrics budget and not impose load to the cluster instance. As a general rule of thumb though we should not restrict operations options.
Since OCP Monitoring (discussed with the team there) allows only to use either cluster monitoring or user-workload monitoring per ns (no mixed cases) we should allow users to setup things according to their needs. By default cluster monitoring is enabled to ship all Source metrics to the cluster Prometheus instance and feed admin dashboards with data (for better UX). 
With this PR there is more control by introducing two vars that affect how much automation we offer:
If `SOURCES_USE_CLUSTER_MONITORING` is set to true cluster monitoring is setup automatically.
If `SOURCES_GENERATE_SERVICE_MONITORS` is set to true service monitors are created for each source.
If user sets cluster monitoring to false and allows the generation of service monitors then automatically the sources are ready to be scraped by the user-workload monitoring Prometheus instance without effort. Enabling user-workload monitoring is a [one step process](https://docs.openshift.com/container-platform/4.7/monitoring/enabling-monitoring-for-user-defined-projects.html).
The user-workload monitoring Prometheus instance will not scrape ns targeted by the cluster one and vice versa, for more check [here](https://github.com/openshift/enhancements/blob/master/enhancements/monitoring/user-workload-monitoring.md) and [here](https://www.openshift.com/blog/thanos-querier-versus-thanos-querier).
These env vars provide the maximum flexibility for the end-user.
- Each time an env update is done at the S-O side, the S-O deployment is restarted and a full resync of source deployments is being done (ctr runtime cache is synched). The same happens if the config for metrics backend is changed at the Eventing CR level. This guarantees that we dont have some sources being updated and some others not. Should be equivalent to the global resync in knative controllers.

skonto commented:
@aliok @markusthoemmes  pls review.

/assign @aliok

skonto commented:
/retest

skonto commented on knative-operator/pkg/monitoring/sources/source_deployment_discovery_controller.go:
```diff
@@ -69,38 +77,102 @@ type ReconcileSourceDeployment struct {
+		}
+		return parsed
+	}
+	return true
```
TODO: remove defaults

skonto commented:
@markusthoemmes gentle ping.

skonto commented:
CSV fields are meant to be idempotent (meant to be metadata shipped together with the operator pod), however we can configure env vars via the subscription as defined [here](https://github.com/operator-framework/operator-lifecycle-manager/blob/master/doc/design/subscription-config.md#configuring-operators-deployed-by-olm). I already tried this via the ODC and it works as expected and we can override any env var defined in the S-O csv.

devguyio commented on knative-operator/pkg/monitoring/sources/source_deployment_discovery_controller.go:
```diff
@@ -69,38 +77,94 @@ type ReconcileSourceDeployment struct {
-func (skipDeletePredicate) Delete(e event.DeleteEvent) bool {
-	return false
+func (r *ReconcileSourceDeployment) reconcileSourceMonitoring(rclient client.Client, dep *v1.Deployment, shouldEnableMonitoring bool, inDeletion bool) error {
+	// Setup monitoring resources only if monitoring should be on and source deployment is not being deleted
```
I'm wondering if we should extract some logic out of this method into other methods to explain the logic more and not rely on the detailed comments. Specially the the two branches of the big if statements? don't have a very strong opinion, just thinking out loud

skonto replied:
Let me think about it.

devguyio reviewed (commented):
left a few small comments, other than that it's lgtm for me. I didn't do a deep of the tests because of the time and limited capacity but that should be ok.
/lgtm

skonto commented on knative-operator/pkg/monitoring/sources/source_deployment_discovery_controller.go:
```diff
@@ -69,38 +77,94 @@ type ReconcileSourceDeployment struct {
+		log.Info("Source in deletion phase")
+		inDeletion = true
+	} else if err != nil {
+		// Do nothing if can't get the source for some other error
```
I think we should catch the error. I will update.

skonto commented:
@devguyio thanks for the review, I addressed the comments. If you have time to test manually then you can set env vars like `USE_CLUSTER_MONITORING_FOR_SOURCES_BY_DEFAULT` at the subscritpion level. 
@rhuss pls review.

skonto commented:
@devguyio gentle ping

rhuss commented on knative-operator/cmd/manager/main.go:
```diff
@@ -118,7 +118,7 @@ func main() {
 	hookServer.Register("/validate-knativekafkas", &webhook.Admission{Handler: knativekafka.NewValidator(mgr.GetClient(), decoder)})
 
-	if err := setupMonitoring(cfg); err != nil {
+	if err := setupServerlesOperatorMonitoring(cfg); err != nil {
```
```suggestion
	if err := setupServerlessOperatorMonitoring(cfg); err != nil {
```

and the corresponding definition

rhuss replied:
Why is "ServerlessOperator" needed here as extra context information ? Is it not clear that this is about the setting up monitoring for Knative/Serverless ?

skonto replied:
I added this for clarity because there are other funcs with similar functionality.

rhuss commented on knative-operator/pkg/monitoring/sources/source_deployment_discovery_controller.go:
```diff
@@ -2,23 +2,31 @@ package sources
 
-var log = common.Log.WithName("source-deployment-discovery-controller")
+var (
+	generateSourceServiceMonitorsEnvVar = "GENERATE_SERVICE_MONITORS_FOR_SOURCES_BY_DEFAULT"
```
Why the extension `BY_DEFAULT` ? How can this be overridden ? (e.g. a user sets this variable, but for certain sources it does not want a ServiceMonitor to be created ?

If this variable triggers to _always_ create a ServiceMonitor for a Source, then I would ommit the `BY_DEFAULT` suffix.

skonto replied:
This can be overridden via a subscription update check comments above. I can ommit BY_DEFAULT.

rhuss commented on knative-operator/pkg/monitoring/sources/source_deployment_discovery_controller.go:
```diff
@@ -2,23 +2,31 @@ package sources
-var log = common.Log.WithName("source-deployment-discovery-controller")
+var (
+	generateSourceServiceMonitorsEnvVar = "GENERATE_SERVICE_MONITORS_FOR_SOURCES_BY_DEFAULT"
+	useClusterMonitoringEnvVar          = "USE_CLUSTER_MONITORING_FOR_SOURCES_BY_DEFAULT"
```
Similar comment as above. If not needed, ommit the `BY_DEFAULT` suffix.

rhuss replied:
Another recommendation. If you start the env var name with "SOURCES" then its much clearer that those env vars belong together. Also, I wonder whether you do need two env vars.

E.g. you could differentiate on the value

```
SOURCES_SERVICE_MONITOR=cluster
SOURCES_SERVICE_MONITOR=generate
```

That makes the configurations surface smaller and less to remember.

skonto replied:
These are different settings. They affect different functionality pls take a look at the description.

rhuss commented on knative-operator/pkg/monitoring/sources/source_deployment_discovery_controller.go:
```diff
@@ -53,7 +61,7 @@ func add(mgr manager.Manager, r reconcile.Reconciler) error {
 		return nil
 	})
-	err = c.Watch(&source.Kind{Type: &v1.Deployment{}}, handler.EnqueueRequestsFromMapFunc(enqueueRequests), skipDeletePredicate{}, skipUpdatePredicate{})
+	err = c.Watch(&source.Kind{Type: &v1.Deployment{}}, handler.EnqueueRequestsFromMapFunc(enqueueRequests))
```
not really deep into it, but why are the predicates not needed anymore ?

skonto replied:
Update is needed when deployments are updated due to changes in the observability config in the eventing CR for example which maps to changes in the deployments env vars. Delete can be safely ignored.

rhuss commented on knative-operator/pkg/monitoring/sources/source_deployment_discovery_controller.go:
```diff
@@ -69,38 +77,119 @@ type ReconcileSourceDeployment struct {
-	if err := monitoring.SetupMonitoringRequirements(r.client, dep); err != nil {
+	err := r.client.Get(context.TODO(), request.NamespacedName, dep)
+	inDeletion := false
+	if apierrors.IsNotFound(err) {
```
Is 'in deletion' the only state that can cause a 404 ?

skonto replied:
Afaik this us typical, the first step in reconciliation is to retrieve the object. Either the call fails for some error, object does not exist (due to deletion) or it exists and then you can update its state etc.

markusthoemmes replied:
From below: I think I follow this now. I feel like it's unnecessary to distinguish between "inDeletion" and "disabling monitoring" though. The code paths would likely benefit from handling both cases the same, I believe.

rhuss reviewed (commented):
@skonto quick review on the UI surface, not deep enough into the weeds to understand the full thing (and the original requirement). Sorry, still time constrained right now.

skonto commented:
I will rename the env vars.

skonto commented:
/retest

skonto commented:
@rhuss ready for another round.

skonto commented:
Validation passes locally not sure why yq fails.

skonto commented:
/retest

skonto commented:
@rhuss gentle ping. Pls take a look.

skonto commented:
@devguyio @rhuss gentle ping

rhuss commented:
yes, sorry. quite busy and many folks are/were on PTO. Hope to continue it still before the weekend, but don't consider me to be a blocker.

skonto commented:
@rhuss we ce can continue on Monday @markusthoemmes will be back, but I think you are the only one who can approve it right now.

skonto commented:
/retest

skonto commented:
/retest

skonto commented:
@markusthoemmes gentle ping.

markusthoemmes commented on knative-operator/pkg/monitoring/sources/source_deployment_discovery_controller.go:
```diff
@@ -69,38 +79,133 @@ type ReconcileSourceDeployment struct {
+		if err := monitoring.RemoveClusterMonitoringRequirements(r.client, dep); err != nil {
+			return err
+		}
+		// No need to do anything if in deletion phase as owner references will do the cleanup
+		if !inDeletion {
```
Ahhhh, you distinguish between "disable monitoring" and "removing the deployment". Is that distinction necessary? I think we can safely drop the `inDeletion` state and handle both cases the same. It's not harmful to actively delete things even if the ownerRef would technically take care of it.

skonto replied:
If I get an update event for a source deployment and monitoring is enabled then I want to avoid setting up the source resources (if in the meanwhile someone deleted the source). Note that when I am saying source resources it is not just the service monitors that have owner refs set to the source but also global rbac rules for the source namespace to be used by Prometheus sa that lives in the openshift-monitoring ns. If I setup source monitoring resources and the source is in deletion phase rbac rules will be left undeleted if that is the only source in that namespace. 
That is why I keep the state there because I have the delete predicate that ignores deletes. I can change that to handle deletions but the logic so far was only to handle updates.

markusthoemmes commented on templates/csv.yaml:
```diff
@@ -467,8 +467,10 @@ spec:
                     - name: DASHBOARDS_ROOT_MANIFEST_PATH
                       value: "deploy/resources/dashboards"
-      
-      # Openshift Ingress adapter for Knative's Ingress implementation.
```
Please don't remove this.

skonto replied:
ok probably removed by accident will bring it back.

skonto commented on knative-operator/pkg/monitoring/sources/source_deployment_discovery_controller.go:
```diff
@@ -69,38 +79,133 @@ type ReconcileSourceDeployment struct {
 	}
-	if err := SetupSourceServiceMonitorResources(r.client, dep); err != nil {
+	eventing := &v1alpha1.KnativeEventing{}
+	if err = r.client.Get(context.TODO(), types.NamespacedName{Namespace: "knative-eventing", Name: "knative-eventing"}, eventing); err != nil {
 		return reconcile.Result{}, err
 	}
-	return reconcile.Result{}, nil
+	shouldEnableMonitoring := okomon.ShouldEnableMonitoring(eventing.Spec.GetConfig())
+	return reconcile.Result{}, r.reconcileSourceMonitoring(dep, shouldEnableMonitoring, inDeletion)
```
If someone disables monitoring eg. set backend to `none` this triggers an update to all source adapter deployments.

skonto replied:
> Shouldn't inDeletion be triggered by a finalizer being present? If we need to cleanup arbitrary (non ownerRef) stuff from the Deployment, we should be adding a finalizer to it. That would cause this to no longer work though, as it'd stick around until the finalizer has been removed by us.

If we were to manage the non ownerRefresources like Prometheus rbac rules yes. Currently if someone accidentally deletes them then I dont restore them unless some source update happens. Otherwise will have to setup a watcher/reconciler for any different kind involved.

skonto commented on knative-operator/pkg/monitoring/sources/source_deployment_discovery_controller.go:
```diff
@@ -69,38 +79,133 @@ type ReconcileSourceDeployment struct {
-type skipDeletePredicate struct {
-	predicate.Funcs
+func (r *ReconcileSourceDeployment) reconcileSourceMonitoring(dep *v1.Deployment, shouldEnableMonitoring bool, inDeletion bool) error {
+	// If monitoring is set to on/off this triggers a global resync to source adapters.
```
Yes.

skonto commented on knative-operator/pkg/monitoring/sources/source_deployment_discovery_controller.go:
```diff
@@ -69,38 +79,133 @@ type ReconcileSourceDeployment struct {
 
-type skipDeletePredicate struct {
-	predicate.Funcs
+func (r *ReconcileSourceDeployment) reconcileSourceMonitoring(dep *v1.Deployment, shouldEnableMonitoring bool, inDeletion bool) error {
```
I changed it because previous review suggested to separate.

skonto commented on knative-operator/pkg/monitoring/sources/source_deployment_discovery_controller.go:
```diff
@@ -69,38 +79,133 @@ type ReconcileSourceDeployment struct {
 	client client.Client
 	scheme *runtime.Scheme
+
+	testShouldUseClusterMonitoring          *bool
+	testShouldGenerateSourceServiceMonitors *bool
```
ok I can set/unset the env vars in tests as an alternative.

skonto commented:
@markusthoemmes I refactored all major issues. I will address the last pending issue which is rbac resource reconciling (right now they are not attached to some owner) in a separate PR because it requires some thought. However, this PR is an improvement in many ways.

skonto commented:
/assign @matzew

matzew reviewed (approved):
/lgtm
/approve

skonto commented:
/hold

skonto commented:
Checking if we constantly trigger a reconciliation.

skonto commented:
@matzew or @rhuss  could you approve once more?

skonto commented:
/unhold

skonto commented:
@matzew gentle ping.

matzew reviewed (approved):
/lgtm
/approve
